{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This is the analysis of gut intestinal stem cell data as presented in BÃ¶ttcher et al. We use in total **75,012** cells from **14 samples** in **6 conditions** (2 replicates in each mutant, 3 replicates in each control and 1 sample FVR only controls):\n",
    "\n",
    "* Control (whole crypt)\n",
    "* Control (FVR enriched)\n",
    "* Control (FVR only)\n",
    "* Mutant (Fltp deficient, whole crypt)\n",
    "* Mutant (Fltp deficient, FVR enriched)\n",
    "* Control (High fat diet project)\n",
    "\n",
    "In this notebook, we load the raw data and filter them according to the thresholds defined by CellRanger.\n",
    "\n",
    "We filter cells by\n",
    "\n",
    "* number of reads\n",
    "* number of detected genes\n",
    "* fraction of mitochondrial reads (< 10%) \n",
    "\n",
    "\n",
    "## Tasks in this notebook\n",
    "\n",
    "In order to keep the notebook structure short and well-arranged, we have split the gut intestinal stem cell analysis into several parts:\n",
    "\n",
    "1. Basic pre-processing and cell filtering\n",
    "2. Preparation of Batch effect correction\n",
    "3. Data analysis with Graph abstraction\n",
    "\n",
    "This notebook covers **part 1**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Scanpy 1.3.1 on 2021-01-22 09:55.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as pl\n",
    "import scanpy.api as sc\n",
    "import matplotlib as mpl\n",
    "from matplotlib import rcParams\n",
    "import scipy as sci\n",
    "\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "sc.logging.print_version_and_date()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.settings.writedir='./../data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def qc_metrics(adata):\n",
    "    \n",
    "    # mean expression per gene\n",
    "    adata.var[\"mean\"] = adata.X.mean(axis=0).T\n",
    "    # fraction of 0s per genes\n",
    "    adata.var[\"frac_zeros\"] = (adata.X.shape[0] - (adata.X > 0).sum(axis=0).T) / adata.X.shape[0]\n",
    "    \n",
    "    # for each cell compute fraction of counts in mito genes vs. all genes\n",
    "    mito_genes = [name for name in adata.var_names if name.startswith('mt-')]\n",
    "    mt_gene_index = np.where(mito_genes)[0]\n",
    "    adata.obs['percent_mito'] = adata[:, mt_gene_index].X.sum(axis=1) / adata.X.sum(axis=1)\n",
    "    # for each cell compute fraction of counts in ribosomal genes vs. all genes\n",
    "    ribo_genes = [name for name in adata.var_names if name.startswith('Rp')]\n",
    "    ribo_gene_index = np.where(ribo_genes)[0]\n",
    "    adata.obs['percent_ribo'] = adata[:, ribo_gene_index].X.sum(axis=1) / adata.X.sum(axis=1)\n",
    "\n",
    "    # total counts and genes per cell\n",
    "    adata.obs['n_counts'] = np.sum(adata.X, axis=1).A1\n",
    "    adata.obs['n_genes'] = np.sum(adata.X > 0, axis=1).A1\n",
    "    print(len(adata.obs_names))\n",
    "    \n",
    "    return(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cell_filter(adata, countmin):\n",
    "    #filter cells with > countmin number of counts and >1000 genes. Countmin is defined based on the CellRanger output. \n",
    "    \n",
    "    # get cell barcode indices\n",
    "    res_cells = sc.pp.filter_cells(adata.X,min_counts=countmin,copy=True)\n",
    "    res_genes = sc.pp.filter_cells(adata.X,min_genes=1000,copy=True)\n",
    "    res_all = np.array([any(tup) for tup in zip(res_cells[0], res_genes[0])])\n",
    "    \n",
    "    # filter\n",
    "    adata_filt=adata.copy()[res_all,:]\n",
    "    print(adata_filt.n_obs)\n",
    "    return(adata_filt)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WT, FVR and mutant mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... reading from cache file ./cache/filtered_data-Control_1-matrix.h5ad\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... reading from cache file ./cache/raw_data-Control_1-matrix.h5ad\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Control_1 = './../filtered_data/Control_1/matrix.mtx'\n",
    "filename_genes_Control_1 = './../filtered_data/Control_1/genes.tsv'\n",
    "filename_barcodes_Control_1 = './../filtered_data/Control_1/barcodes.tsv'\n",
    "\n",
    "adata_Control_1 = sc.read(filename_data_Control_1, cache=True).transpose()\n",
    "adata_Control_1.var_names = np.genfromtxt(filename_genes_Control_1, dtype=str)[:, 1]\n",
    "adata_Control_1.obs_names = np.genfromtxt(filename_barcodes_Control_1, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Control_1 = './../raw_data/Control_1/matrix.mtx'\n",
    "filename_genes_Control_1 = './../raw_data/Control_1/genes.tsv'\n",
    "filename_barcodes_Control_1 = './../raw_data/Control_1/barcodes.tsv'\n",
    "\n",
    "adata_Control_1_raw = sc.read(filename_data_Control_1, cache=True).transpose()\n",
    "adata_Control_1_raw.var_names = np.genfromtxt(filename_genes_Control_1, dtype=str)[:, 1]\n",
    "adata_Control_1_raw.obs_names = np.genfromtxt(filename_barcodes_Control_1, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3440\n",
      "filtered out 733840 cells that have less than 6604.0 counts\n",
      "filtered out 733248 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "4047\n"
     ]
    }
   ],
   "source": [
    "adata_Control_1 = qc_metrics(adata_Control_1)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Control_1.obs['n_counts'])\n",
    "del adata_Control_1\n",
    "adata_Control_1 = cell_filter(adata_Control_1_raw, countmin)\n",
    "del adata_Control_1_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Control_2 = './../filtered_data/Mutant_1/matrix.mtx'   #####two samples were exchanged \n",
    "filename_genes_Control_2 = './../filtered_data/Mutant_1/genes.tsv'\n",
    "filename_barcodes_Control_2 = './../filtered_data/Mutant_1/barcodes.tsv'\n",
    "\n",
    "adata_Control_2 = sc.read(filename_data_Control_2, cache=True).transpose()\n",
    "adata_Control_2.var_names = np.genfromtxt(filename_genes_Control_2, dtype=str)[:, 1]\n",
    "adata_Control_2.obs_names = np.genfromtxt(filename_barcodes_Control_2, dtype=str)\n",
    "\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Control_2 = './../raw_data/Mutant_1/matrix.mtx'   #####two samples were exchanged \n",
    "filename_genes_Control_2 = './../raw_data/Mutant_1/genes.tsv'\n",
    "filename_barcodes_Control_2 = './../raw_data/Mutant_1/barcodes.tsv'\n",
    "\n",
    "adata_Control_2_raw = sc.read(filename_data_Control_2).transpose()\n",
    "adata_Control_2_raw.var_names = np.genfromtxt(filename_genes_Control_2, dtype=str)[:, 1]\n",
    "adata_Control_2_raw.obs_names = np.genfromtxt(filename_barcodes_Control_2, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2940\n",
      "filtered out 734340 cells that have less than 5691.0 counts\n",
      "filtered out 733749 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "3542\n"
     ]
    }
   ],
   "source": [
    "adata_Control_2 = qc_metrics(adata_Control_2)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Control_2.obs['n_counts'])\n",
    "del adata_Control_2\n",
    "adata_Control_2 = cell_filter(adata_Control_2_raw, countmin)\n",
    "del adata_Control_2_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    },
    "run_control": {
     "marked": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Control_3_FVR = './../filtered_data/Control_3_FVR/matrix.mtx'\n",
    "filename_genes_Control_3_FVR = './../filtered_data/Control_3_FVR/genes.tsv'\n",
    "filename_barcodes_Control_3_FVR = './../filtered_data/Control_3_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Control_3_FVR = sc.read(filename_data_Control_3_FVR, cache=True).transpose()\n",
    "adata_Control_3_FVR.var_names = np.genfromtxt(filename_genes_Control_3_FVR, dtype=str)[:, 1]\n",
    "adata_Control_3_FVR.obs_names = np.genfromtxt(filename_barcodes_Control_3_FVR, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Control_3_FVR = './../raw_data/Control_3_FVR/matrix.mtx'\n",
    "filename_genes_Control_3_FVR = './../raw_data/Control_3_FVR/genes.tsv'\n",
    "filename_barcodes_Control_3_FVR = './../raw_data/Control_3_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Control_3_FVR_raw = sc.read(filename_data_Control_3_FVR, cache=True).transpose()\n",
    "adata_Control_3_FVR_raw.var_names = np.genfromtxt(filename_genes_Control_3_FVR, dtype=str)[:, 1]\n",
    "adata_Control_3_FVR_raw.obs_names = np.genfromtxt(filename_barcodes_Control_3_FVR, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3766\n",
      "filtered out 733514 cells that have less than 10409.0 counts\n",
      "filtered out 732917 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "4416\n"
     ]
    }
   ],
   "source": [
    "adata_Control_3_FVR = qc_metrics(adata_Control_3_FVR)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Control_3_FVR.obs['n_counts'])\n",
    "del adata_Control_3_FVR\n",
    "adata_Control_3_FVR = cell_filter(adata_Control_3_FVR_raw, countmin)\n",
    "del adata_Control_3_FVR_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Control_4_FVR = './../filtered_data/Mutant_3_FVR/matrix.mtx'  #####two samples were exchanged \n",
    "filename_genes_Control_4_FVR = './../filtered_data/Mutant_3_FVR/genes.tsv'\n",
    "filename_barcodes_Control_4_FVR = './../filtered_data/Mutant_3_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Control_4_FVR = sc.read(filename_data_Control_4_FVR, cache=True).transpose()\n",
    "adata_Control_4_FVR.var_names = np.genfromtxt(filename_genes_Control_4_FVR, dtype=str)[:, 1]\n",
    "adata_Control_4_FVR.obs_names = np.genfromtxt(filename_barcodes_Control_4_FVR, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Control_4_FVR = './../raw_data/Mutant_3_FVR/matrix.mtx'  #####two samples were exchanged \n",
    "filename_genes_Control_4_FVR = './../raw_data/Mutant_3_FVR/genes.tsv'\n",
    "filename_barcodes_Control_4_FVR = './../raw_data//Mutant_3_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Control_4_FVR_raw = sc.read(filename_data_Control_4_FVR, cache=True).transpose()\n",
    "adata_Control_4_FVR_raw.var_names = np.genfromtxt(filename_genes_Control_4_FVR, dtype=str)[:, 1]\n",
    "adata_Control_4_FVR_raw.obs_names = np.genfromtxt(filename_barcodes_Control_4_FVR, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4566\n",
      "filtered out 732714 cells that have less than 8212.0 counts\n",
      "filtered out 731954 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "5338\n"
     ]
    }
   ],
   "source": [
    "adata_Control_4_FVR = qc_metrics(adata_Control_4_FVR)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Control_4_FVR.obs['n_counts'])\n",
    "del adata_Control_4_FVR\n",
    "adata_Control_4_FVR = cell_filter(adata_Control_4_FVR_raw,countmin)\n",
    "del adata_Control_4_FVR_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Mutant_1 = './../filtered_data/Control_2/matrix.mtx' #####two samples were exchanged \n",
    "filename_genes_Mutant_1 = './../filtered_data/Control_2/genes.tsv'\n",
    "filename_barcodes_Mutant_1 = './../filtered_data/Control_2/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_1 = sc.read(filename_data_Mutant_1, cache=True).transpose()\n",
    "adata_Mutant_1.var_names = np.genfromtxt(filename_genes_Mutant_1, dtype=str)[:, 1]\n",
    "adata_Mutant_1.obs_names = np.genfromtxt(filename_barcodes_Mutant_1, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Mutant_1 = './../raw_data/Control_2/matrix.mtx' #####two samples were exchanged \n",
    "filename_genes_Mutant_1 = './../raw_data/Control_2/genes.tsv'\n",
    "filename_barcodes_Mutant_1 = './../raw_data/Control_2/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_1_raw = sc.read(filename_data_Mutant_1, cache=False).transpose()\n",
    "adata_Mutant_1_raw.var_names = np.genfromtxt(filename_genes_Mutant_1, dtype=str)[:, 1]\n",
    "adata_Mutant_1_raw.obs_names = np.genfromtxt(filename_barcodes_Mutant_1, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3386\n",
      "filtered out 733894 cells that have less than 8673.0 counts\n",
      "filtered out 733346 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "3949\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_Mutant_1 = qc_metrics(adata_Mutant_1)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Mutant_1.obs['n_counts'])\n",
    "del adata_Mutant_1\n",
    "adata_Mutant_1 = cell_filter(adata_Mutant_1_raw, countmin)\n",
    "del adata_Mutant_1_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Mutant_2 = './../filtered_data/Mutant_2/matrix.mtx'   \n",
    "filename_genes_Mutant_2 = './../filtered_data/Mutant_2/genes.tsv'\n",
    "filename_barcodes_Mutant_2 = './../filtered_data/Mutant_2/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_2 = sc.read(filename_data_Mutant_2, cache=True).transpose()\n",
    "adata_Mutant_2.var_names = np.genfromtxt(filename_genes_Mutant_2, dtype=str)[:, 1]\n",
    "adata_Mutant_2.obs_names = np.genfromtxt(filename_barcodes_Mutant_2, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Mutant_2 = './../raw_data/Mutant_2/matrix.mtx'   \n",
    "filename_genes_Mutant_2 = './../raw_data/Mutant_2/genes.tsv'\n",
    "filename_barcodes_Mutant_2 = './../raw_data/Mutant_2/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_2_raw = sc.read(filename_data_Mutant_2, cache=True).transpose()\n",
    "adata_Mutant_2_raw.var_names = np.genfromtxt(filename_genes_Mutant_2, dtype=str)[:, 1]\n",
    "adata_Mutant_2_raw.obs_names = np.genfromtxt(filename_barcodes_Mutant_2, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3611\n",
      "filtered out 733669 cells that have less than 7461.0 counts\n",
      "filtered out 733119 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "4167\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_Mutant_2 = qc_metrics(adata_Mutant_2)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Mutant_2.obs['n_counts'])\n",
    "del adata_Mutant_2\n",
    "adata_Mutant_2 = cell_filter(adata_Mutant_2_raw, countmin)\n",
    "del adata_Mutant_2_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Mutant_3_FVR = './../filtered_data/Control_4_FVR/matrix.mtx'  #####two samples were exchanged \n",
    "filename_genes_Mutant_3_FVR = './../filtered_data/Control_4_FVR/genes.tsv'\n",
    "filename_barcodes_Mutant_3_FVR = './../filtered_data/Control_4_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_3_FVR = sc.read(filename_data_Mutant_3_FVR, cache=True).transpose()\n",
    "adata_Mutant_3_FVR.var_names = np.genfromtxt(filename_genes_Mutant_3_FVR, dtype=str)[:, 1]\n",
    "adata_Mutant_3_FVR.obs_names = np.genfromtxt(filename_barcodes_Mutant_3_FVR, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Mutant_3_FVR = './../raw_data/Control_4_FVR/matrix.mtx'  #####two samples were exchanged \n",
    "filename_genes_Mutant_3_FVR = './../raw_data/Control_4_FVR/genes.tsv'\n",
    "filename_barcodes_Mutant_3_FVR = './../raw_data/Control_4_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_3_FVR_raw = sc.read(filename_data_Mutant_3_FVR, cache=True).transpose()\n",
    "adata_Mutant_3_FVR_raw.var_names = np.genfromtxt(filename_genes_Mutant_3_FVR, dtype=str)[:, 1]\n",
    "adata_Mutant_3_FVR_raw.obs_names = np.genfromtxt(filename_barcodes_Mutant_3_FVR, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3153\n",
      "filtered out 734127 cells that have less than 11376.0 counts\n",
      "filtered out 733307 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "3978\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_Mutant_3_FVR = qc_metrics(adata_Mutant_3_FVR)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Mutant_3_FVR.obs['n_counts'])\n",
    "del adata_Mutant_3_FVR\n",
    "adata_Mutant_3_FVR = cell_filter(adata_Mutant_3_FVR_raw,countmin)\n",
    "del adata_Mutant_3_FVR_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_Mutant_4_FVR = './../filtered_data/Mutant_4_FVR/matrix.mtx'\n",
    "filename_genes_Mutant_4_FVR = './../filtered_data/Mutant_4_FVR/genes.tsv'\n",
    "filename_barcodes_Mutant_4_FVR = './../filtered_data/Mutant_4_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_4_FVR = sc.read(filename_data_Mutant_4_FVR, cache=True).transpose()\n",
    "adata_Mutant_4_FVR.var_names = np.genfromtxt(filename_genes_Mutant_4_FVR, dtype=str)[:, 1]\n",
    "adata_Mutant_4_FVR.obs_names = np.genfromtxt(filename_barcodes_Mutant_4_FVR, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_Mutant_4_FVR = './../raw_data/Mutant_4_FVR/matrix.mtx'\n",
    "filename_genes_Mutant_4_FVR = './../raw_data/Mutant_4_FVR/genes.tsv'\n",
    "filename_barcodes_Mutant_4_FVR = './../raw_data/Mutant_4_FVR/barcodes.tsv'\n",
    "\n",
    "adata_Mutant_4_FVR_raw = sc.read(filename_data_Mutant_4_FVR, cache=True).transpose()\n",
    "adata_Mutant_4_FVR_raw.var_names = np.genfromtxt(filename_genes_Mutant_4_FVR, dtype=str)[:, 1]\n",
    "adata_Mutant_4_FVR_raw.obs_names = np.genfromtxt(filename_barcodes_Mutant_4_FVR, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3561\n",
      "filtered out 733719 cells that have less than 12290.0 counts\n",
      "filtered out 732677 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "4608\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_Mutant_4_FVR = qc_metrics(adata_Mutant_4_FVR)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_Mutant_4_FVR.obs['n_counts'])\n",
    "del adata_Mutant_4_FVR\n",
    "adata_Mutant_4_FVR = cell_filter(adata_Mutant_4_FVR_raw, countmin)\n",
    "del adata_Mutant_4_FVR_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Maren.Buettner/miniconda3/lib/python3.6/site-packages/ipykernel_launcher.py:8: DeprecationWarning: Use obs_names instead of smp_names, smp_names will be removed in the future.\n",
      "  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_FVR = './../filtered_data/Control_6/matrix.mtx'\n",
    "filename_genes_FVR = './../filtered_data/Control_6/genes.tsv'\n",
    "filename_barcodes_FVR = './../filtered_data/Control_6/barcodes.tsv'\n",
    "\n",
    "adata_FVR = sc.read(filename_data_FVR, cache=True).transpose()\n",
    "adata_FVR.var_names = np.genfromtxt(filename_genes_FVR, dtype=str)[:, 1]\n",
    "adata_FVR.smp_names = np.genfromtxt(filename_barcodes_FVR, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_FVR = './../raw_data/Control_6/matrix.mtx'\n",
    "filename_genes_FVR = './../raw_data/Control_6/genes.tsv'\n",
    "filename_barcodes_FVR = './../raw_data/Control_6/barcodes.tsv'\n",
    "\n",
    "adata_FVR_raw = sc.read(filename_data_FVR, cache=True).transpose()\n",
    "adata_FVR_raw.var_names = np.genfromtxt(filename_genes_FVR, dtype=str)[:, 1]\n",
    "adata_FVR_raw.obs_names = np.genfromtxt(filename_barcodes_FVR, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6041\n",
      "filtered out 731239 cells that have less than 8784.0 counts\n",
      "filtered out 729720 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "7574\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_FVR = qc_metrics(adata_FVR)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_FVR.obs['n_counts'])\n",
    "del adata_FVR\n",
    "adata_Control_6 = cell_filter(adata_FVR_raw, countmin)\n",
    "del adata_FVR_raw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comment: `Control_7_FVR_only` contains about twice as much cells as expected. The reason is that the fastq-files of both `Control_5_FVR` and `Control_7_FVR` have been aggregated into one file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_FVR = './../filtered_data/Control_5_FVR/matrix.mtx'\n",
    "filename_genes_FVR = './../filtered_data/Control_5_FVR/genes.tsv'\n",
    "filename_barcodes_FVR = './../filtered_data/Control_5_FVR/barcodes.tsv'\n",
    "\n",
    "adata_FVR = sc.read(filename_data_FVR, cache=True).transpose()\n",
    "adata_FVR.var_names = np.genfromtxt(filename_genes_FVR, dtype=str)[:, 1]\n",
    "adata_FVR.obs_names = np.genfromtxt(filename_barcodes_FVR, dtype=str)\n",
    "\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_FVR = './../raw_data/Control_5_FVR/matrix.mtx'\n",
    "filename_genes_FVR = './../raw_data/Control_5_FVR/genes.tsv'\n",
    "filename_barcodes_FVR = './../raw_data/Control_5_FVR/barcodes.tsv'\n",
    "\n",
    "adata_FVR_raw = sc.read(filename_data_FVR, cache=True).transpose()\n",
    "adata_FVR_raw.var_names = np.genfromtxt(filename_genes_FVR, dtype=str)[:, 1]\n",
    "adata_FVR_raw.obs_names = np.genfromtxt(filename_barcodes_FVR, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7802\n",
      "filtered out 729478 cells that have less than 7078.0 counts\n",
      "filtered out 729397 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "8906\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_FVR = qc_metrics(adata_FVR)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_FVR.obs['n_counts'])\n",
    "del adata_FVR\n",
    "adata_Control_5_FVR = cell_filter(adata_FVR_raw,countmin)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_FVR_only = './../filtered_data/Control_7_FVR_only/matrix.mtx'\n",
    "filename_genes_FVR_only = './../filtered_data/Control_7_FVR_only/genes.tsv'\n",
    "filename_barcodes_FVR_only = './../filtered_data/Control_7_FVR_only/barcodes.tsv'\n",
    "\n",
    "adata_FVR_only = sc.read(filename_data_FVR_only, cache=True).transpose()\n",
    "adata_FVR_only.var_names = np.genfromtxt(filename_genes_FVR_only, dtype=str)[:, 1]\n",
    "adata_FVR_only.obs_names = np.genfromtxt(filename_barcodes_FVR_only, dtype=str)\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_FVR_only = './../raw_data/Control_7_FVR_only/matrix.mtx'\n",
    "filename_genes_FVR_only = './../raw_data/Control_7_FVR_only/genes.tsv'\n",
    "filename_barcodes_FVR_only = './../raw_data/Control_7_FVR_only/barcodes.tsv'\n",
    "\n",
    "adata_FVR_only_raw = sc.read(filename_data_FVR_only, cache=True).transpose()\n",
    "adata_FVR_only_raw.var_names = np.genfromtxt(filename_genes_FVR_only, dtype=str)[:, 1]\n",
    "adata_FVR_only_raw.obs_names = np.genfromtxt(filename_barcodes_FVR_only, dtype=str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15948\n",
      "filtered out 721332 cells that have less than 7298.0 counts\n",
      "filtered out 724800 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "17333\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_FVR_only = qc_metrics(adata_FVR_only)\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_FVR_only.obs['n_counts'])\n",
    "del adata_FVR_only\n",
    "adata_Control_7_FVR_only = cell_filter(adata_FVR_only_raw,countmin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7298.0\n"
     ]
    }
   ],
   "source": [
    "print(countmin)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check the number of duplicated barcodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes_list = {}\n",
    "barcodes_list['Control_5_FVR'] = adata_Control_5_FVR.obs_names.values\n",
    "barcodes_list['Control_7_FVR_only'] = adata_Control_7_FVR_only.obs_names.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes_dup = np.intersect1d(barcodes_list['Control_5_FVR'], barcodes_list['Control_7_FVR_only'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8898"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(barcodes_dup)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot library size relation of those two samples (using the duplicate barcodes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "adata_Control_5_FVR.obs['n_counts'] = np.sum(adata_Control_5_FVR.X, axis=1).A1\n",
    "adata_Control_7_FVR_only.obs['n_counts'] = np.sum(adata_Control_7_FVR_only.X, axis=1).A1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZwAAAEXCAYAAACZNvIiAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xu4XFWZ7/vvLysLWCAQEKQhXIKKeEAUJErsuD2IFxBvNF7ARkTalt7dagu6sZOGPqCCxM0WFe+gKChCAHGJgqaRgH2aI9GkE4hBIgGBZIGAhiANQULynj/mqKRSqao1Z92r1u/zPOtJ1Ziz5hxzTah3jTHeOYYiAjMzs3ab1O0KmJnZxOCAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAYxOepPskvb5L595N0n9IekLS57pw/vdL+s9On9cmpsndroDZBHcK8Edgh6jyUJyk7wCrIuLMPAcrun+vkTQN+D0wHBHPdrc21mpu4Zi1iKRG/oDbB7izWrAxGzQOONaTUjfX/5J0h6THJc2VtE3atkU3kKSQ9ML0+juSvirpp5L+W9Ktkv5K0hckPSbpLkmHVJzyFZLuTNu/XTpXOt5bJC2RtEbS/yfppRX1/BdJdwBPVgs6kv5a0q/Tdfxa0l+X6gmcBHwi1fP1FZ87BTihbPuPU/n/JemWVJ9lkt42zv6zJN2Tuu3ulPQ3Be7Dq9M1r5G0UtL7U/mOki6T9Kik+yWdKWlS2na2pO+VHWNauj+T0/tbJH063ZcnJP27pF3S7v+R/l2TruFVkl4o6Rfp9/dHSXPz1t96TET4xz899wPcB/wK2APYGfgt8D/TtvcD/1mxfwAvTK+/Q9ZNdSiwDTCfrJvmfcAQcA5wc8W5fgPslc51K3BO2nYI8AhwWPrsSWn/rcs+uyR9dqTKdewMPAacSNaF/Z70/rlldT2nzu9hs+3AMLAC+FdgK+AI4Alg/1rHA96Vfo+TgOOAJ4Hda/0uyz63Tzr2e9J5nwscnLZdBvwI2B6YBvwO+EDadjbwvbLjTEv3Z3J6fwtwD/AiYCS9n1Nt31R2BXBGqv82wKu7/d+nfxr7cQvHetmFEfFgRKwGfgwcXOCzP4yIRRHxNPBD4OmIuCwi1gNzyQJJuS9HxMp0rnPJvmQhG2P5RkQsiIj1EXEp8BdgRkU9V0bE2ir1eDNwd0R8NyKejYgrgLuAtxa4lnIzgOeQfUE/ExHzgZ+U1XcLEXF1+j1uiIi5wN3AK3Oc62+Bn0fEFRGxLiL+FBFLJA0BxwOzI+KJiLgP+BxZUM3r2xHxu/Q7u4r693YdWfDbIyKejggnOfQpBxzrZX8oe/0U2RdtXg+XvV5b5X3lsVaWvb6frEUA2Rfdx1OX0hpJa8haM3vU+GylPdLxyt0PTK1f/brHWxkRG/IeT9L7yroE1wAvAXaptX+ZvchaIpV2IWvxlF9X0Wsqcm8/AQj4VepC/LsC57Ee4oBj/ehJYNvSG0l/1YJj7lX2em/gwfR6JXBuREwp+9k2tVRK6g34P0gWtMrtDYzlrFflsR8E9iqNl1Q53mb7S9oHuBj4MFk33hSy7kPlOPdK4AVVyv/IplZHtTpsdn+AIvdni99lRPwhIj4YEXsA/wB8tTReZ/3FAcf60e3AgZIOToP7Z7fgmB+StKekncnGC0oD0xcD/1PSYcpsJ+nNkrbPedwbgBdJ+ltJkyUdBxxA1g2Wx8PA88veLyBrEXxC0rCkw8m6566ssf92ZF/ijwJIOpmshZPH5cDrJb071f25kg5O3ZJXAedK2j4FtY8BpUSBJcBrJO0taUdgds7zkeq5ofwaJL1L0p7p7WPpejZU+az1OAcc6zsR8TvgU8DPycYjWtGn/33g34F7ybqRzknnWgh8EPgy2ZfdCrKB9rx1/RPwFuDjwJ/IuofeEhF/zHmIbwEHpO6w0Yh4hizAvImspfFV4H0RcVeN/e8kG1/5JVkwOogsKSJP3R8Ajk51X00WSF6WNn+ErCVzL9nv//vAJelzN5IF7DuAReQPrkTEU2RjaLema5gBvAJYIOm/geuAj0bEvXmPab1DEU7/NzOz9nMLx8zMOsIBx8zMOsIBx8zMOsIBx8zMOsKzRZfZZZddYtq0ad2uhplZX1m0aNEfI2LX8fZzwCkzbdo0Fi5c2O1qmJn1FUmVs2lU5S41MzPrCAccMzPrCAccMzPrCAccMzPrCAccMzPrCGepmZlNUKOLxzh/3nIeXLOWPaaMcPqR+3PMIY0u1TQ+BxwzswlodPEYs69dytp16wEYW7OW2dcuBWhb0HGXmpnZBHT+vOUbg03J2nXrOX/e8rad0y0cM7MJpNSNNrZmbdXtD9YobwUHHDOzCaKyG62aPaaMtO387lIzM5sgqnWjlRsZHuL0I/dv2/ndwjEzmyDqdZdNdZaamZm1yh5TRqqO3UydMsKts45o+/ndpWZmNkGcfuT+jAwPbVbW7m60cm7hmJlNEKXusk4+7FnOAcfMbAI55pCpHQswldylZmZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHeGAY2ZmHdHWgCPpEkmPSPpNWdnOkm6UdHf6d6dULkkXSloh6Q5JLy/7zElp/7slnVRWfqikpekzF0pSvXOYmXXT6OIxZs6Zz76zrmfmnPmMLh7rdpU6qt0tnO8AR1WUzQJuioj9gJvSe4A3Afuln1OAr0EWPICzgMOAVwJnlQWQrwEfLPvcUeOcw8ys40YXj3HAv/2UU+cuYWzNWgIYW7OW2dcunVBBp60BJyL+A1hdUfx24NL0+lLgmLLyyyJzGzBF0u7AkcCNEbE6Ih4DbgSOStt2iIjbIiKAyyqOVe0cZmYdNbp4jNOvvp2n1m3YYtvades5f97yLtSqO7oxhrNbRDyUXv8B2C29ngqsLNtvVSqrV76qSnm9c5iZddT585azbkPU3P7gmrUdrE13dTVpILVMat+JDpxD0imSFkpa+Oijj7azKmY2AY0XUPaYMtKhmnRfNwLOw6k7jPTvI6l8DNirbL89U1m98j2rlNc7xxYi4qKImB4R03fdddeGL8rMrJp6AUXA6Ufu37nKdFk3As51QCnT7CTgR2Xl70vZajOAx1O32DzgjZJ2SskCbwTmpW1/ljQjZae9r+JY1c5hZtZRpx+5P8OTVHXbCTP25phDplbdNogmt/Pgkq4ADgd2kbSKLNtsDnCVpA8A9wPvTrvfABwNrACeAk4GiIjVkj4N/Drt96mIKCUi/BNZJtwI8NP0Q51zmJl1VCmgnH3dMtasXQfATtsOc9ZbD5xQwQZA2RCHAUyfPj0WLlzY7WqYmfUVSYsiYvp4+3mmATMz6wgHHDMz6wgHHDMz6wgHHDMz6wgHHDMz64i2pkWbmfWj0cVjnD9vOQ+uWcseU0Y4/cj9J1wKczs44JiZlRldPMbsa5eydt16YNOszoCDTpMccMzM2NSqGasy91lpVmcHnOY44JjZhHfm6FIuv+2BujMJT6RZndvFSQNmNqGNLh4bN9jAxJrVuV0ccMxsQjt/3vJxg83I8NCEmtW5XdylZmYT2nhdZVOdpdYyDjhmNiHUSnXeY8pI1UQBAZ8/7mAHmhZywDGzgVeZFFCe6nz6kftvlgYNWbCZaGvVdIIDjpkNtFpJAaVU51tnHQHgBz07wAHHzAZavaSA0vjNMYdMdYDpAGepmdlAq5cU4FTnznILx8xaphfmIKusw5Rth3nsqXVb7CdwqnOHOeCYWUt0cw6yM0eXcvmCB4iKvrOxNWsZniSGh8S69Zs2OimgOxxwzAZYJ1sc589bvlmmF7R/DrLRxWP867V38NS6DTX3WbchmDIyzHZbT3ZSQJc54JgNqE63OGqNlbRrDrJN11c72JQ8vnYdS856Y1vqYfk5acBsQNVrcbRDrQH4dg3MV7u+Wpwc0BsccMwGVKdbHKcfuT8jw0OblbV6DrLRxWPMnDOffWddX3V2gGo8D1rvyN2lJula4FvATyNi/DasmXVVrSlb2vXXfqmbrh1jRqOLxzj7umWsWbtltlk9O207zFlvPdDjNT2iyBjOV4GTgQslXQ18OyLa0zY3s6ZVm7Kl3X/tt+MByjNHl/K92x4o9JlthyfxmWNf6kDTY3IHnIj4OfBzSTsC70mvVwIXA9+LiGJ/epi1US88D9Jt7WxxdEreYCPoy+ubaAplqUl6LvBe4ERgMXA58GrgJODwVlfOrBFek36TfpyypWiLZuqUkY3zoVlvy500IOmHwP8LbAu8NSLeFhFzI+IjwHOKnljSaZKWSfqNpCskbSNpX0kLJK2QNFfSVmnfrdP7FWn7tLLjzE7lyyUdWVZ+VCpbIWlW0fpZ/+p0dlarlA+Iz5wzn9HFY92uUscVDTZOCOgvRVo4F0bEzdU2RMT0IieVNBX4Z+CAiFgr6SrgeOBo4PMRcaWkrwMfAL6W/n0sIl4o6Xjgs8Bxkg5InzsQ2IOsm+9F6TRfAd4ArAJ+Lem6iLizSD2tP3U6O6sVJnqrrJGkgEmC8449aEL8fgbFuAFH0rHVXpdExLVNnHtE0jqyVtNDwBHA36btlwJnkwWct6fXANcAX5akVH5lRPwF+L2kFcAr034rIuLeVO8r074OOBNAp7OzWqEbT+l3W2mcLW96c7mhSeJz73rZwP5uBlWeFs5b62wLoHDAiYgxSf8HeABYC/w7sAhYExHPpt1WAaX/mqYCK9Nnn5X0OPDcVH5b2aHLP7OyovywanWRdApwCsDee+9d9FKsy6olB3QjO6tZ/dgqa0Zli66I7bYa4ty/ccumH40bcCLi5FafVNJOZC2OfYE1wNXAUa0+Tx4RcRFwEcD06dNrLZthPahWN9R5xx7Eecce1FfZWf3YKmtGkVkCSrYaEv/7nW7V9LMiD35uDbwDmFb+uYj4VAPnfT3w+4h4NB37WmAmMEXS5NTK2RMojZqOAXsBqyRNBnYE/lRWXlL+mVrlNiDqdUPdOuuIvvpi6sdWWSNOuPiX3HrP6sKf2+9523Hjxw5vfYWso4okDfwIeJys6+svTZ73AWCGpG3JutReBywEbgbeCVxJlmr9o7T/den9L9P2+RERkq4Dvi/pArKkgf2AX5Gl5e8naV+yQHM8m8aGbEAMUjdUK56Z6fVnjw4790YefuKZQp/xTAGDpUjA2TMiWtLtFRELJF0D/BfwLNkzPRcB1wNXSjonlX0rfeRbwHdTUsBqsgBCRCxLGW53puN8KCLWA0j6MDAPGAIuiYhlrai79Y5B64Zq5pmZXs5yy7OEQDXvnbE35xxzUJtqZd2gqFyxqNaO0kXAlyJiaXur1D3Tp0+PhQsXdrsallO1geeR4aGBSZUt0mKZOWd+1eDbzYcis/tzR67lA8qNDE/iPE9L01ckLcrzeEyRFs6rgfdL+j1Zl5qAiIiXNlhHs6YMwtQttRRtsfRa92Ij858BfOG4gwfi/ll1RQLOm9pWC7MG9ePULXkUfS6nU92L47W6Gk0KgKwLbRDvpW2Se2qbiLgfmEL2XM5bgSmpzKxv9Mv0MUVbLJ1ai2b2tUsZW7OWYFOrq/Q7fMMFtzQVbDxeM/iKpEV/FPggmx70/J6kiyLiS22pmVmL9fLAeqWiLZZOdC/WanWdfd0yPnHN7TyzvthjbH6Ac+Ip0qX2AeCwiHgSQNJnydKUHXCsLzQ7fUwn044beS6n3d2LtVpXRRdFc2tm4ioScASU/9+6PpWZ9YVmBtY73TpqpMXS7oBYq9VVxH7P287BZgIrEnC+DSxIyxQAHMOm52TMel4zA+vdmFyzSIulEwGxWqurCM8WYEWSBi4gW2J6dfo5OSK+UNqe5kcz61nNDKz3WtpxpU6sAbTw/tUNBRuRpTs72FihFT8j4r/IZgeo5ibg5U3XyAZSL0y70szAeq/PatDOgNjoMzXg8RrbXKGAMw6P5/SgXvii76XssEYH1nt9cs12BcQ3XHALdz/yZOHPzXzBzlz+wVc1dW4bPLm71HLw1P49ZrznJjqlX5d8LnfMIVM579iDmDplBJFNGdNLU+i0+jmc0cVjTJt1fUPBZr/nbedgY1W1soVjPaZXVpHs9fGPvHp5VoNWPofTzGwBbtlYPe5SG2C98kXf6+Mfg6IVAXHarOsLf8bZZ5ZXw11qkqZIOqOs6HUtqI+1UK0v9E5/0Xdi2pWJpB3T85S60IraYeshBxvLbdwWjqS9gH8jW+BsFLgC+BRwYnoNQEQ01ga3tumVge5uzOrcC8kS7dDqBIxmus/csrGi8nSpXQb8AvgBcBTZypxLgJdGxB/aWDdrUi9N39/J8Y9eyoprtVaOy730rJ/x578Uf67GgcYalSfg7BwRZ6fX8yS9CzghIoqtqmRd0csD3e3SK8kS7dCqcbkXn3EDTxecbBP8XI01J1fSQJpFoJQU8CdgR0kCd6VZ7+mVZIl2aDYBY99Z1zf0/MIOWw9xxydbssK8TWB5As6ObDm7QOl9AM9vaY3Mcqo1TjPIWXGNjsuNLh7j1LlLGjqnU52tVcYNOBExrQP1sD7WjQH6euM07U6W6GZCQiPjco12n3msxlotT5bancDlwBURcW/7q2T9pFsD9PXGaW6ddcTGfVodFHohISHvuNxh597Iw08809A5vnDcwX0/3mW9J0+X2nuA44EbJf2JLBV6bkQ82NaaWV/o1gD9eOM07UqW6IeEhGa6z7YZEnede3SLa2SWydOldjtwOzBb0gzgOOA2SfcA34+Ii9tcRxtHN7t4ujVA361xml5PSGi0VeNAY51QaKaBiLgtIk4D3gdMAb7cllpZbt2eoLNbsxl0a/aCXpm9oVJppoBGgs3MF+zsYGMdkXsuNUmvIOteewfwe+AbwNVtqtfAq9UqKdpa6XYXT6sG6Ited7ceam3metvVEm001Rngvjlvbvr8ZnnlSRr4DFk32mrgSmBmRKxqd8UGWa2B54X3r+YHi8YKDUh3u4unFV/8jQ7Ed+Oh1kav98zRpVx+2wMbA0Orkg0amf8MnIFm3aGI+n8bSfp/yDLU7m7piaUpwDeBl5A9z/N3wHJgLjANuA94d0Q8lh4y/SJwNPAU8P60+iiSTgLOTIc9JyIuTeWHAt8BRoAbgI/GOBc7ffr0WLhwYesusoaZc+ZXHX8YklhfpYpTp4xszLzKe6x6n+k19X4fGyL6fi600cVjnDZ3SdVWSKP36YWzr+fZBps1zkCzVpO0KCKmj7dfnjGcbUrBRtIbmq7ZJl8EfhYRLwZeBvwWmAXcFBH7kS1ZPSvt+yZgv/RzCvC1VJ+dgbOAw4BXAmelWRFI+3yw7HM985h0rdZHtWBTb38YjJmY6/0+urlwXKucP295zS6voi3Rl571M6bNaizY7Pe87bhvzpsdbKxr8gSc8i/qz7bipJJ2BF4DfAsgIp6JiDXA24FL026XAsek128HLovMbcAUSbsDRwI3RsTqiHgMuBE4Km3bISU5BNkEpKVjdV2tAeYhVV9SqN6AdK+vRJlHngH3flshtFy9oFIk2WDarOsbmmxzh62HuG/Om92FZl3XrRU/9wUeBb4t6WXAIuCjwG4R8VDa5w/Abun1VGBl2edXpbJ65auqlG9B0ilkrSb23nvvxq+ogFoDz+84dOpmYzil8vFaK/0+QWe130c1vZJ6XFStFG5BrpaoH+C0QZEn4DxP0sfI/v8ovd4oIi5o8LwvBz4SEQskfZFN3Wel44akRpNvcouIi4CLIBvDaff5oP7A8/R9du6J5QQ6qfL3ManGWFae1kAvroNTLaAKOGHG3nXr1swDnOAMNOs9eQLOxcD2VV43YxWwKiIWpPfXkAWchyXtHhEPpW6xR9L2MWCvss/vmcrGgMMrym9J5XtW2b9n1GqV9HtrpVHl112ZtQb5J6js9rQz1TSS2faGC27h7keebOh8XkLAelWemQY+medAkmZHxHl59o2IP0haKWn/iFhOtjz1nennJGBO+vdH6SPXAR+WdCVZgsDjKSjNAz5TlijwRmB2RKyW9Oc0M8ICsgdVv5SnbtZ9jaYed/uZpHry/iHhVo0NslaO4bwLyBVwko8Al0vaCrgXOJksieEqSR8A7gfenfa9gSwlegVZWvTJkK3FI+nTwK/Tfp8qW5/nn9iUFv3T9GM5dbtrqjLolBIG6tWh288kNavRFTjBz9VYf2hlwKmeYlVDRCwBquVtv67KvgF8qMZxLgEuqVK+kOwZHyuoF7qmGqlDv66D00yrZrJgxXlu1Vh/KDSX2jg6MuBu7Veva6qX69DOZ5JGF48xc8589p11PTPnzG/ZM0HTZl3fcLB574y9HWysr3SthWO9qxVdU812ydU619iatYwuHquZcAGtn1+tHS2+Ey7+Jbfe0/jq7B6rsX7UyoDjiTwHRLNdU9W+oE+du4Szr1vG2W87MNeXdK06AHW/7NuR5dfqZIRG5z8DL/ds/W3cLjVJF0iaOd5+EfGZ1lTJuq3RrqlSt9Opc5dUfYhzzdp1uaeoqVaHkk5377UqGeGwc29sONhMVtaqcbCxfpanhXMi8BpJu5JNrHlFRCxub7Wsncbr7mqka6raszPV5G0ZlLbXGt/oZOZZK5IR3KoxyxdwVkXEdEkvIlum4HuShsiWmr4iIn7X1hpaS+UdjyjaNVWt26mWvMHimEOmcv685Q1/2bcqtbuZNXCamZbGq3DaoMmTpRYAEfG7iPh0RBxI9nzMNmTPx1iXFcmgalcGWpEWR5GWQTPde61aCbXRCVIbXYETsu4zBxsbNHlaOFtkn0XEHcAdwOyW18gKKZpB1a6HI+sN8pcrmqbcK7MOFGnxNdN95sk2bZDlCTj/I8+BJB0YEcuarI8VVPSLtV0PR9abAfvmux5tqluryLQwpcBUb/2Zds6i0EywcaqzDbo8c6n9d85jfZdsBmjroKItlmbGI+pp1zMweeVNWthxZLgtsyg0swKnkwJsovCDn32uaIul0Qy0PPt3c6brPEkLI8NDSLS0q82TbZrlp6ixrHHhA0n/FRF93cKZPn16LFy4sNvVKKTWVP6tWvWz2vFFlkkytUfWmwHYd9b1NbvRBBsD5Wlzl1TdT8Dnjzs4dyBuNtB4sk0bJJIWRUS1uTE3060VP61FarVYAGbOmd9091a1lkPpC7tX1puB2i29qVNGuHXWERvf10qznrJt/q62ZrrPwK0am7haGXAay/+0plV2ZeXJXMvbTTZe9lqvrDeTd2yq1n4R+branBRg1rhxA46kut1kEfFf6d8ZraqUNWe8zLUiqdR50p17Yb2ZvGNTtfY7bZwZDZpZgVPA7x1szHK1cD5XZ1sAR9TZbl0wXuZakVTqai2CSr2y3kzepIVq+9Wb0cCtGrPWyJMW/dpOVMRqK/rcyHiZa0VSqctbBGNr1m5MGChp1XoztXRq5dFagTXPw6zVeGE0sy3lHsORNAz8I/CaVHQL8I2IWNeGelnSyFos441nNJJKXWvs57Uv3pXz5y3ntLlLWh4QOrnyaGVgbYZnCzCrLndatKRvAsPApanoRGB9RPx9m+rWcZ1Oi87z1/vMOfNzZV8VOXarUqnzHKeZFkqj196MZrrPwF1oNjG1Iy36FRHxsrL38yXdXrxqBvn/em907rN64xmtmhWglckJ5UpBqlZLo5kkhVoB8MzRpXzvtgcaPq4Djdn4igSc9ZJeEBH3AEh6PpBvPnrbQt6B+3bNfdaKWQFamZxQkmeKmkavvVYAbOYBTmegmeVXJOCcDtws6V6y/8/2AU5uS60mgLwtl3bNfdYKtYLhjiPDQGOts/GmqGnm2msFwEa5VWNWTK6AI2kSsBbYDyj93748Iv7SrooNovLunEkS66uMn1X+9d7tSTHrOf3I/Tn96ttZt2Hz63jymWcZXTyWq3VW2cVVb8C+2al0WvW80G7bb8WCM97QkmOZTSS5Ak5EbJD0lYg4hGwdHCuosjunWrCp9dd7NyfFrOeYQ6byyR8v47GnNk9UXLc+OH/e8nFbZ9W6uCrTrktakSiQd82eetyqMWtcnhU/S26S9A5JnhW6AbW6ioakQqtI9po1T1XPin9wzdpxV8qsNU9b5X9grehCHF081lSw2WHrIQcbsyYVGcP5B+BjwLOSniZNGhwRO7SlZgOmVnfOhojNBp079aBjq4zXbVavdVbrd1KaibpVv4NmpqUBt2rMWiV3wImI7Vt9cklDwEJgLCLeImlf4ErgucAi4MSIeEbS1sBlwKHAn4DjIuK+dIzZwAfIMub+OSLmpfKjgC8CQ8A3I2JOq+tfRN7xjKJpxN0OUM0kNeSd4blRzS4hAA42Zq2Uu0tN0k15ygr6KPDbsvefBT4fES8EHiMLJKR/H0vln0/7IekA4HjgQOAo4KuShlIg+wrwJuAA4D1p3645/cj9GRke2qys8ou5XhpxNaUANZaWVC4FqNHFYy2vfy3jdZvVk+d30ojRxWNMm3V90wujOdiYtVae2aK3AbYFdpG0E5u62HcAGv5TWtKewJuBc4GPpbGhI4C/TbtcCpwNfA14e3oNcA3w5bT/24ErU7bc7yWtAF6Z9lsREfemc12Z9r2z0fo2K0+2WdE04kaec2mHPEkN9VpirWyhnXDxL7n1ntUNf/69M/bmnGMOavjzZlZbni61fwBOBfYg6+YqBZw/A19u4txfAD4BlLrqngusiYhn0/tVbApoU4GVABHxrKTH0/5TgdvKjln+mZUV5Yc1UdeWGO+LuehDno3OQtBp43UVtio4eloas96WZ7boLwJflPSRiPhSK04q6S3AIxGxSNLhrThmE3U5BTgFYO+99+5mVQqPh7RrFoJWa1dLrLzV1MxC6Q40Zp1RJGngS5L+GphW/rmIuKyB884E3ibpaGAbsu65LwJTJE1OrZw9gdJgxBiwF7BK0mRgR7LkgVJ5SflnapVXXtdFwEWQTd7ZwLW0zDGHTGXh/au5YsFK1kcwJPGOQ2u3APIGqG4nFrSyJVY+z1qtZ3aKcLAx65wiyxN8F3gBsIRNc6gFWfZYIRExG5idjns48L8i4gRJVwPvJMtUOwn4UfrIden9L9P2+RERkq4Dvi+P5cGcAAAUCklEQVTpArIuv/2AX5F1++2Xst7GyBILSmNDPevM0aVcftsDG79E10fwg0VjTN9n56oBIs8YSCen+K+lVS2xymtxq8asvxR5Dmc6cEDkXc+gMf8CXCnpHGAx8K1U/i3guykpYDVZACEilkm6iiwZ4FngQxGxHkDSh4F5ZGnRl0TEsjbWG2iuJTG6eGyzYFMyXtfTeGMgvZBY0Kr54MabZy0vBxuz7igScH4D/BXwUCsrEBG3kC3mRsoqe2WVfZ4G3lXj8+eSZbpVlt8A3NDCqtbVbEvi/HnLa/7F3kwSQC8kFpSuv3wanK0nF5nkIuNpacz6W5GAswtwp6RfARsn7YyIt7W8Vn2o2ZZEvQDQTBJA0e6sRltpeT739LoNG1+vWbuuUEB2BppZ/ysScM5uVyUGQa2AMbZmLfvOun7cL+8dR4ZZs7b6vGR5u56qfekX6c6q1ko7be4SFt6/uu6zKXlad42ujfOJa27nmfWN9+I60Jj1jtz9GhHxC+Ausudmtgd+m8qM+q2Q8WYBGF08xp+frh5sth2elLuFUW3WASD3TAC1JtO8/LYH6s5ekGeGhKJde6VpaRxszAZHkSy1dwPnk423CPiSpNMj4po21a2vVGtJVKr2F30pUGyo8b26tqwbqp56X/q3zjqiqW69SMcvOglneXmRrj1Ptmk2mIp0qZ0BvCIiHgGQtCvwc7KpZia8yhTlvAkA42Ve5R2/aUVyQL31YsYbYxovmORZG6f0fE0zHGzMeleRgDOpFGySP1FsPZ2BV56iPHPO/Fx/0df7Ii99IecZkG/Fsy6nH7k/p81dUjVY1jtOnnGies8MeVZns4mhSMD5maR5wBXp/XF0MO243+T9i75WS2hI4rxjs4H6POnW1ZZ7Hp6kQs+6lGY6qHweaLxnZvJOwlntmaEzR5fyvdseyF3HSp5s06x/5Jkt+oXAbhFxuqRjgVenTb8ELm9n5frZeH/Rjzfes8PI5I2fz53dVblUZgNrs55zzEFM32fnwqnRjUzC6VRns4lF400cIOknwOyIWFpRfhDwmYh4axvr11HTp0+PhQsXtv08tbrbKo0MD9UMSoLNVgqtdcxWLWbWSs12oTnQmPUWSYsiYvp4++XpUtutMtgARMRSSdMaqNuEl3cgf+269QxJrK/yR8GOI8PMnDN/YyukkcH+Vsr7wKhbNWYTV56AM6XOtt6aB79P1HvIs9L6iC1aOsOTxJPPPLvxGPVmTu7EUgV5Hvx0qrOZ5ckyWyjpg5WFkv6ebEE2K0gFxlZKD2qWP7i51eRJrKt4IDLYcsimPMtt5pz57DvrembOmd/yJajHe/Bz2qzrHWzMLFcL51Tgh5JOYFOAmQ5sBfxNuyo2yNY8la91U8oyKx+Qrzf+EWQBqbxbC/JluTWjVnfe2Jq1TXWhOdCYDZY8K34+DPy1pNcCL0nF10fE/LbWbIDVG3Mp95xtJm8RFMqni6lULUFg5pz5VVsfH7/qdk6bu6TpBdlGF4+1ZCG0Sg42ZoOnyIqfNwM3t7EuE0aeaXCgekuoXhJAtWdlau1fSkRotsVT71miRjjQmA0uzxTQBcccMnWzcZmhGoM61Qb8ayUBTBkZrhow8iQNVE60WUQrs+AcbMwGW5GZBqyFKsdl8i4hUGsGg7PfduBm+5XPTZany6vRwJG3e7AeBxqzicEBpwfknRom776VAayUwRZQ87meRtOnmwk2kwUrznOwMZsoHHB6RJGpYcbbt9a6NlNTcMrbmqrHD3CaWVEOOD2o0WWeS+otVZC3NVWvDs0Em/2etx03fuzwhj9vZv1r3LnUJpJOzaVWy+jiMc6+btkWsxCMDA/VXKWzmmbnVas2pjQ8pC0eNi3KrRqzwZR3LjVnqfWI0pd8tSlvimaRnX7k/owMD21WlrfbbHTxGB+/6vYtuuSaCTb3zXmzg42ZuUutV4y38meRLLIiSQjlSkGvWlJBoxxozKzEAadHjBdQRoa3bIzWG2dpZH2a8YJeEQ40ZlbJAadHjPc8y1PrNjC6eKzmsztja9Zy6twlnH3dMt7yst25+a5Hc7VuyoNWq9o1DjZmVo2TBsq0Immg0QyzPKuAwqbU5tJDnXnUSjrIe868HGjMJqa8SQMOOGWaDTi1ZgyolWFWGZxe++JduWLBynHHUIYniXUbit23ygy10cVjfOyqJRQ8TFXbDIm7zj26+QOZWV/q6Sw1SXtJulnSnZKWSfpoKt9Z0o2S7k7/7pTKJelCSSsk3SHp5WXHOintf7ekk8rKD5W0NH3mQqnIKjSNGW9dmHKl4DSWurLG1qzlB4vGeM9he22xrk2losEGNh8jGl08xunX3N6SYHPfnDc72JhZLt1Ki34W+HhEHADMAD4k6QBgFnBTROwH3JTeA7wJ2C/9nAJ8DbIABZwFHAa8EjirFKTSPh8s+9xR7b6oeg9cVqoVnG6+61FOmLH3uEGnqEnSxoXXzp+3vOlnaibLXWhmVkxXkgYi4iHgofT6CUm/BaYCbwcOT7tdCtwC/Esqvyyy/r/bJE2RtHva98aIWA0g6UbgKEm3ADtExG2p/DLgGOCn7byuWgP/1eYpqxeczjnmIKbvs3OhcZrxrI/YuAyBJ9s0s27o+oOfkqYBhwALgN1SMAL4A7Bbej0VWFn2sVWprF75qirl1c5/iqSFkhY++uijTV1LkQcua02WWSo/5pCp3DrrCHbadrhQHaZOGeG9M/auuuTB2nXra64WmodbNWbWjK6mRUt6DvAD4NSI+HP5MEtEhKS2ZzRExEXARZAlDTRzrCIPXI43iWYpoeCxnMtRw+aJAZff9kAzl7IFBxoza1bXAo6kYbJgc3lEXJuKH5a0e0Q8lLrMHknlY8BeZR/fM5WNsakLrlR+Syrfs8r+bZf3gct6wamRdOXKllQr1qkB+MJxBze8/LSZWbmuBJyUMfYt4LcRcUHZpuuAk4A56d8flZV/WNKVZAkCj6egNA/4TFmiwBuB2RGxWtKfJc0g66p7H/Cltl9YQbWCU9En/qdWtKRGF4/x5F+ebbp+btWYWSt1q4UzEzgRWCqpNKjwr2SB5ipJHwDuB96dtt0AHA2sAJ4CTgZIgeXTwK/Tfp8qJRAA/wR8BxghSxZoa8JAPUUfBi0yb5pgi+drmhmnAQcaM2uPbmWp/SfUzPx9XZX9A/hQjWNdAlxSpXwh8JImqtkS1aagKWWL1Qo6RbrDypMPHGzMrJd5LrU2q/cwaK2AUy2hoJrhSeKpZ55l31nXs83wJNau29BwPR1ozKzdHHDarMjDoCWlQPTJHy+rmaUmwbMbYuN2Bxsz63UOOG023sOgtcZ3Sj+l7WNr1iLYOKNzK6bAc6Axs07q+oOfg67aw6ACXvviXavOpzb72qUbp6CBTQ+ATp0y0rLlA8DBxsw6zy2cNjvmkKksvH81l9/2wKbWCfCDRWNcf8dDucd3WjXFjQONmXWLA04H3HzXo1u0TtauW18zKaBIWnRekwUrznOwMbPuccDpgKIBpHJmZ0+2aWaDwAGnA2olDkwZGeYvz27YoqWzPqLp52lKHGzMrFc4aaADas0iffbbDuS8Yw9iUhuWhrtvzpsdbMysp7iF0wHjzSLdqtYMeLJNM+tdDjgdkncW6Wa4RWNmvcwBZwA40JhZP3DA6aITLv4lt96zevwd63CwMbN+4YDTBWeOLuV7Ta7I6UBjZv3GAadDyudEa5aDjZn1IwecNqo18WajHGjMrJ854LRJ5cJrDjZmNtE54LTJJ3+8bNwF1PJwoDGzQeGZBtpgdPFYzYXTinCwMbNB4hZOizkDzcysOgecFmr2uZodth7ijk8e1cIamZn1DgecFpk26/qmPu9WjZkNOgecJo0uHmtq8k1PtmlmE4UDThOaDTZu1ZjZROKA04RGg40DjZlNRE6L7jAHGzObqAa6hSPpKOCLwBDwzYiY0626ONCY2UQ3sC0cSUPAV4A3AQcA75F0QDfq4mBjZjbYLZxXAisi4l4ASVcCbwfubNUJdtt+Kx5+4pma2x1ozMw2GdgWDjAVWFn2flUq24ykUyQtlLTw0UcfLXSCBWe8gd2232qL8t2238rBxsyswiC3cHKJiIuAiwCmT59eeFLnBWe8oeV1MjMbRIPcwhkD9ip7v2cqMzOzLhjkgPNrYD9J+0raCjgeuK7LdTIzm7AGtkstIp6V9GFgHlla9CURsazL1TIzm7AGNuAARMQNwA3droeZmQ12l5qZmfUQRRROzBpYkh4F7s+x6y7AH9tcnW4Y1OuCwb22Qb0uGNxrG8Tr2icidh1vJwecBkhaGBHTu12PVhvU64LBvbZBvS4Y3Gsb1OvKw11qZmbWEQ44ZmbWEQ44jbmo2xVok0G9LhjcaxvU64LBvbZBva5xeQzHzMw6wi0cMzPrCAccMzPrCAecAiQdJWm5pBWSZnW7PtVI2kvSzZLulLRM0kdT+c6SbpR0d/p3p1QuSRema7pD0svLjnVS2v9uSSeVlR8qaWn6zIWS1OFrHJK0WNJP0vt9JS1I9Zmb5s5D0tbp/Yq0fVrZMWan8uWSjiwr78o9ljRF0jWS7pL0W0mvGpR7Jum09N/ibyRdIWmbfrxnki6R9Iik35SVtf0e1TpHX4oI/+T4IZuP7R7g+cBWwO3AAd2uV5V67g68PL3eHvgd2Yqn/xuYlcpnAZ9Nr48GfgoImAEsSOU7A/emf3dKr3dK236V9lX67Js6fI0fA74P/CS9vwo4Pr3+OvCP6fU/AV9Pr48H5qbXB6T7tzWwb7qvQ928x8ClwN+n11sBUwbhnpGtQfV7YKTsXr2/H+8Z8Brg5cBvysrafo9qnaMff7pegX75AV4FzCt7PxuY3e165aj3j4A3AMuB3VPZ7sDy9PobwHvK9l+etr8H+EZZ+TdS2e7AXWXlm+3XgevZE7gJOAL4Sfqf84/A5Mr7RDZx66vS68lpP1Xeu9J+3brHwI7pS1kV5X1/z9i0EOLO6R78BDiyX+8ZMI3NA07b71Gtc/Tjj7vU8su1gmgvSd0RhwALgN0i4qG06Q/Abul1reuqV76qSnmnfAH4BLAhvX8usCYinq1Sn43XkLY/nvYves3tti/wKPDt1FX4TUnbMQD3LCLGgP8DPAA8RHYPFtH/96ykE/eo1jn6jgPOgJL0HOAHwKkR8efybZH9qdR3+fCS3gI8EhGLul2XFptM1lXztYg4BHiSrOtkoz6+ZzsBbycLqnsA2wFHdbVSbdKJe9Sv/x2UOODk1zcriEoaJgs2l0fEtan4YUm7p+27A4+k8lrXVa98zyrlnTATeJuk+4ArybrVvghMkVRaaqO8PhuvIW3fEfgTxa+53VYBqyJiQXp/DVkAGoR79nrg9xHxaESsA64lu4/9fs9KOnGPap2j7zjg5NcXK4imzJZvAb+NiAvKNl0HlDJiTiIb2ymVvy9l1cwAHk/N93nAGyXtlP5KfSNZX/lDwJ8lzUjnel/ZsdoqImZHxJ4RMY3s9z8/Ik4AbgbeWePaStf8zrR/pPLjU0bUvsB+ZAO2XbnHEfEHYKWk/VPR64A7GYB7RtaVNkPStuncpWvr63tWphP3qNY5+k+3B5H66Ycs8+R3ZFkxZ3S7PjXq+GqyJvcdwJL0czRZP/hNwN3Az4Gd0/4CvpKuaSkwvexYfwesSD8nl5VPB36TPvNlKga7O3Sdh7MpS+35ZF8+K4Crga1T+Tbp/Yq0/fllnz8j1X85ZRlb3brHwMHAwnTfRskymAbingGfBO5K5/8uWaZZ390z4Aqycah1ZK3SD3TiHtU6Rz/+eGobMzPrCHepmZlZRzjgmJlZRzjgmJlZRzjgmJlZRzjgmJlZRzjgmJlZRzjgmNUh6a8kXSnpHkmLJN0g6UUNHOdUSds28LlbJE0fZ/tySUvSz/Pq7Hu2pLGyfedIOkvSeRX7HSzpt+n1fWnK/Dsk/ULSPkWvwazEAceshvTE9w+BWyLiBRFxKNlsxI1MnngqUDXgSBpqvJYAnBARB6ef8aY9+XzZvrPIHmY8rmKf41N5yWsj4qXALcCZTdbVJjAHHLPaXgusi4ivlwoi4nbgPyWdr2xBsaWSjgOQdHhqcZQWUrs8TW3yz2QTV94s6ea0739L+pyk24FXSXpdmil6qbKFvrbuxAVGxO+AxyQdVlb8bjYPOCW/pMdnSLfe5oBjVttLyKbSr3Qs2VQ0LyObnPL80uSKZMtBnEq2YNjzgZkRcSHwIFlL4bVpv+3IFuV6GdmUNt8BjouIg8hmj/7HAvX8duoi+7fUKqvntLIutdKqmVeQtWpI836tjoi7q3z2KLJpd8wa4oBjVtyrgSsiYn1EPAz8AnhF2variFgVERvI5rGbVuMY68lm9AbYn2xG5d+l95eSrS6ZxwkpSP2P9HPiOPuXd6nNS2VzgXdKmsSW3WmQtczGgDdV2WaWmwOOWW3LgEMLfuYvZa/Xk7VWqnk6ItY3VKsykS1wRkQ8Qbbs9isbOMZKshVH/2/gHWQBqNxrgX3IAugnm6mvTWwOOGa1zQe2lnRKqUDSS4E1wHGShiTtStYa+dU4x3oC2L7GtuXANEkvTO9PJGs11SVpsqRd0uth4C1ksw034grg88C9EbGqcmNkq2+eSjbl/s4NnsMmOAccsxoim0r9b4DXp7ToZcB5ZC2JO4DbyYLSJyJb06aei4CflZIGKs7zNHAycLWkpWTLZ3+9cr8qtgbmSSotRTEGXJzr4rZ0NXAgdbrMIluz5QrgQw2ewyY4L09gZmYd4RaOmZl1RK0BTTPrIZIWkHWhlTsxIpZW2fcM4F0VxVdHxLntqp9ZHu5SMzOzjnCXmpmZdYQDjpmZdYQDjpmZdYQDjpmZdcT/D5oVB7pmHQGQAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "pl.scatter(x=adata_Control_5_FVR[np.in1d(adata_Control_5_FVR.obs_names, barcodes_dup)].obs['n_counts'], \n",
    "           y = adata_Control_7_FVR_only[np.in1d(adata_Control_7_FVR_only.obs_names, barcodes_dup)].obs['n_counts'])\n",
    "pl.title('number of total counts')\n",
    "pl.xlabel('Control_5_FVR')\n",
    "pl.ylabel('Control_7_FVR_only')\n",
    "pl.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We observe that most duplicated barcodes are have exactly the same number of total counts. However, some show a higher total count in `Control_7_FVR_only`, but not in `Control_5_FVR`. This can be explained by an expected random barcode overlap of 1-2% for any two unrelated samples - therefore, cells whose total counts is not exactly the same in both samples, are two different cells with the same barcode.\n",
    "In order to mend the issue that `Control_7_FVR_only` also contains cells from `Control_5_FVR`, we subtract the raw count matrix of `Control_5_FVR` from the raw count matrix of `Control_7_FVR_only`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['AAACCTGAGAAACCAT-1', 'AAACCTGAGAAACCGC-1', 'AAACCTGAGAAACCTA-1',\n",
       "       'AAACCTGAGAAACGAG-1', 'AAACCTGAGAAACGCC-1', 'AAACCTGAGAAAGTGG-1',\n",
       "       'AAACCTGAGAACAACT-1', 'AAACCTGAGAACAATC-1', 'AAACCTGAGAACTCGG-1',\n",
       "       'AAACCTGAGAACTGTA-1',\n",
       "       ...\n",
       "       'TTTGTCATCTTGGGTA-1', 'TTTGTCATCTTGTACT-1', 'TTTGTCATCTTGTATC-1',\n",
       "       'TTTGTCATCTTGTCAT-1', 'TTTGTCATCTTGTTTG-1', 'TTTGTCATCTTTACAC-1',\n",
       "       'TTTGTCATCTTTACGT-1', 'TTTGTCATCTTTAGGG-1', 'TTTGTCATCTTTAGTC-1',\n",
       "       'TTTGTCATCTTTCCTC-1'],\n",
       "      dtype='object', length=737280)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_FVR_only_raw.obs_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "737280"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.intersect1d(adata_FVR_raw.obs_names,adata_FVR_only_raw.obs_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "737280"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_FVR_raw.n_obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  (19, 7)\t1.0\n",
      "  (19, 20)\t1.0\n",
      "  (19, 25)\t1.0\n",
      "  (19, 29)\t2.0\n",
      "  (19, 39)\t1.0\n",
      "  (19, 51)\t7.0\n",
      "  (19, 52)\t1.0\n",
      "  (19, 62)\t6.0\n",
      "  (19, 63)\t1.0\n",
      "  (19, 122)\t2.0\n",
      "  (19, 131)\t1.0\n",
      "  (19, 152)\t1.0\n",
      "  (19, 163)\t1.0\n",
      "  (19, 188)\t2.0\n",
      "  (37, 131)\t1.0\n",
      "  (54, 62)\t1.0\n",
      "  (78, 62)\t1.0\n",
      "  (112, 62)\t1.0\n",
      "  (128, 62)\t1.0\n",
      "  (156, 186)\t1.0\n",
      "  (156, 188)\t1.0\n",
      "  (161, 62)\t1.0\n",
      "  (170, 153)\t1.0\n",
      "  (172, 7)\t1.0\n",
      "  (172, 51)\t2.0\n",
      "  (172, 62)\t4.0\n",
      "  (182, 52)\t1.0\n",
      "  (182, 62)\t1.0\n",
      "  (186, 62)\t1.0\n"
     ]
    }
   ],
   "source": [
    "print(adata_FVR_only_raw.X[:200,:200])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  (19, 7)\t1.0\n",
      "  (19, 20)\t1.0\n",
      "  (19, 25)\t1.0\n",
      "  (19, 29)\t2.0\n",
      "  (19, 39)\t1.0\n",
      "  (19, 51)\t7.0\n",
      "  (19, 52)\t1.0\n",
      "  (19, 62)\t6.0\n",
      "  (19, 63)\t1.0\n",
      "  (19, 122)\t2.0\n",
      "  (19, 131)\t1.0\n",
      "  (19, 152)\t1.0\n",
      "  (19, 163)\t1.0\n",
      "  (19, 188)\t2.0\n",
      "  (37, 131)\t1.0\n",
      "  (78, 62)\t1.0\n",
      "  (112, 62)\t1.0\n",
      "  (128, 62)\t1.0\n",
      "  (156, 188)\t1.0\n",
      "  (161, 62)\t1.0\n",
      "  (170, 153)\t1.0\n",
      "  (172, 7)\t1.0\n",
      "  (172, 51)\t2.0\n",
      "  (172, 62)\t4.0\n",
      "  (182, 52)\t1.0\n",
      "  (182, 62)\t1.0\n"
     ]
    }
   ],
   "source": [
    "print(adata_FVR_raw.X[:200,:200])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "adata_tmp = adata_FVR_only_raw.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tmp.X = adata_FVR_only_raw.X - adata_FVR_raw.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã n_vars = 737280 Ã 27998 "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  (54, 62)\t1.0\n",
      "  (156, 186)\t1.0\n",
      "  (186, 62)\t1.0\n"
     ]
    }
   ],
   "source": [
    "print(adata_tmp.X[:200,:200])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check whether single values are lower than 0. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  (13, 27876)\tTrue\n",
      "  (19, 15076)\tTrue\n",
      "  (23, 12264)\tTrue\n",
      "  (261, 227)\tTrue\n",
      "  (261, 7876)\tTrue\n",
      "  (618, 15524)\tTrue\n",
      "  (796, 14935)\tTrue\n",
      "  (913, 12343)\tTrue\n",
      "  (913, 15516)\tTrue\n",
      "  (913, 16497)\tTrue\n",
      "  (927, 15525)\tTrue\n",
      "  (962, 27908)\tTrue\n",
      "  (1122, 25106)\tTrue\n",
      "  (1150, 15524)\tTrue\n",
      "  (1150, 17128)\tTrue\n",
      "  (1197, 11509)\tTrue\n",
      "  (1197, 15503)\tTrue\n",
      "  (1304, 15525)\tTrue\n",
      "  (1314, 15525)\tTrue\n",
      "  (1368, 15551)\tTrue\n",
      "  (1387, 19164)\tTrue\n",
      "  (1490, 15520)\tTrue\n",
      "  (1576, 7422)\tTrue\n",
      "  (1576, 15524)\tTrue\n",
      "  (1891, 15525)\tTrue\n",
      "  :\t:\n",
      "  (735962, 27228)\tTrue\n",
      "  (735985, 15525)\tTrue\n",
      "  (736105, 7557)\tTrue\n",
      "  (736147, 22932)\tTrue\n",
      "  (736191, 15525)\tTrue\n",
      "  (736276, 15525)\tTrue\n",
      "  (736431, 18103)\tTrue\n",
      "  (736465, 15525)\tTrue\n",
      "  (736517, 15504)\tTrue\n",
      "  (736596, 470)\tTrue\n",
      "  (736596, 5645)\tTrue\n",
      "  (736596, 11830)\tTrue\n",
      "  (736596, 12297)\tTrue\n",
      "  (736596, 18104)\tTrue\n",
      "  (736596, 24968)\tTrue\n",
      "  (736596, 24976)\tTrue\n",
      "  (736596, 27914)\tTrue\n",
      "  (736608, 14817)\tTrue\n",
      "  (736648, 13207)\tTrue\n",
      "  (736648, 27911)\tTrue\n",
      "  (736738, 15525)\tTrue\n",
      "  (736822, 22493)\tTrue\n",
      "  (736822, 23346)\tTrue\n",
      "  (736872, 25686)\tTrue\n",
      "  (736986, 1422)\tTrue\n"
     ]
    }
   ],
   "source": [
    "print(adata_tmp.X<0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_FVR_only_raw.X[13, 27876]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_FVR_raw.X[13, 27876]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CellRanger did not seem to aggregate all reads from the fastq files in the exact same way."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1.0     18320\n",
       "-2.0       408\n",
       "-3.0        85\n",
       "-4.0        38\n",
       "-5.0        23\n",
       "-6.0        14\n",
       "-7.0         7\n",
       "-8.0         7\n",
       "-11.0        6\n",
       "-9.0         4\n",
       "-10.0        3\n",
       "-15.0        3\n",
       "-13.0        2\n",
       "-19.0        2\n",
       "-20.0        1\n",
       "-12.0        1\n",
       "-22.0        1\n",
       "-57.0        1\n",
       "-17.0        1\n",
       "-18.0        1\n",
       "-33.0        1\n",
       "-40.0        1\n",
       "-21.0        1\n",
       "-60.0        1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.value_counts(adata_tmp.X[adata_tmp.X<0].A1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The difference is interesting, though we could not detect the source of it. We set all negative values to 0. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_tmp.X[adata_tmp.X<0] = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filter `Control_7_FVR_only` based on `countmin` value determined by CellRanger. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 729007 cells that have less than 7298.0 counts\n",
      "filtered out 732682 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "8511\n"
     ]
    }
   ],
   "source": [
    "adata_Control_7_FVR_only = cell_filter(adata_tmp,countmin)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make `var_names` unique in all anndata objects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_Control_1.var_names_make_unique()\n",
    "adata_Control_2.var_names_make_unique() \n",
    "adata_Control_3_FVR.var_names_make_unique() \n",
    "adata_Control_4_FVR.var_names_make_unique()\n",
    "adata_Mutant_1.var_names_make_unique() \n",
    "adata_Mutant_2.var_names_make_unique() \n",
    "adata_Mutant_3_FVR.var_names_make_unique() \n",
    "adata_Mutant_4_FVR.var_names_make_unique()\n",
    "adata_Control_5_FVR.var_names_make_unique() \n",
    "adata_Control_6.var_names_make_unique() \n",
    "adata_Control_7_FVR_only.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concatenate all samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/Maren.Buettner/miniconda3/lib/python3.6/site-packages/anndata/base.py:1806: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  obs = pd.concat(out_obss)\n"
     ]
    }
   ],
   "source": [
    "adata_all=adata_Control_1.concatenate([adata_Control_2, adata_Control_3_FVR, adata_Control_4_FVR,\n",
    "                                       adata_Mutant_1, adata_Mutant_2, adata_Mutant_3_FVR, adata_Mutant_4_FVR,\n",
    "                                       adata_Control_5_FVR, adata_Control_6, adata_Control_7_FVR_only],\n",
    "                                       batch_key='sample',\n",
    "                                       batch_categories=['Control_1', 'Control_2', 'Control_3_FVR', 'Control_4_FVR',\n",
    "                                                         'Mutant_1', 'Mutant_2', 'Mutant_3_FVR', 'Mutant_4_FVR',\n",
    "                                                         'Control_5_FVR', 'Control_6', 'Control_7_FVR_only'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 11585 genes that are detected in less than 20 cells\n"
     ]
    }
   ],
   "source": [
    "sc.pp.filter_genes(adata_all, min_cells=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "59036\n"
     ]
    }
   ],
   "source": [
    "adata_all=qc_metrics(adata_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã n_vars = 59036 Ã 16413 \n",
       "    obs: 'n_counts', 'sample', 'percent_mito', 'percent_ribo', 'n_genes'\n",
       "    var: 'n_cells', 'mean', 'frac_zeros'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Control_5_FVR         8906\n",
       "Control_7_FVR_only    8511\n",
       "Control_6             7574\n",
       "Control_4_FVR         5338\n",
       "Mutant_4_FVR          4608\n",
       "Control_3_FVR         4416\n",
       "Mutant_2              4167\n",
       "Control_1             4047\n",
       "Mutant_3_FVR          3978\n",
       "Mutant_1              3949\n",
       "Control_2             3542\n",
       "Name: sample, dtype: int64"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_all.obs['sample'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_all = adata_all[adata_all.obs['percent_mito'] < 0.1, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.write('./../data/new_filter_all.h5', adata_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FOXA2 mice "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_CD_1 = './../filtered_data/CD_1/matrix.mtx'\n",
    "filename_genes_CD_1 = './../filtered_data/CD_1/genes.tsv'\n",
    "filename_barcodes_CD_1 = './../filtered_data/CD_1/barcodes.tsv'\n",
    "\n",
    "adata_CD_1 = sc.read(filename_data_CD_1, cache=True).transpose()\n",
    "adata_CD_1.var_names = np.genfromtxt(filename_genes_CD_1, dtype=str)[:, 1]\n",
    "adata_CD_1.obs_names = np.genfromtxt(filename_barcodes_CD_1, dtype=str)\n",
    "\n",
    "adata_CD_1.var_names_make_unique\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_CD_1 = './../raw_data/CD_1/matrix.mtx'\n",
    "filename_genes_CD_1 = './../raw_data/CD_1/genes.tsv'\n",
    "filename_barcodes_CD_1 = './../raw_data/CD_1/barcodes.tsv'\n",
    "\n",
    "adata_CD_1_raw = sc.read(filename_data_CD_1, cache=True).transpose()\n",
    "adata_CD_1_raw.var_names = np.genfromtxt(filename_genes_CD_1, dtype=str)[:, 1]\n",
    "adata_CD_1_raw.obs_names = np.genfromtxt(filename_barcodes_CD_1, dtype=str)\n",
    "\n",
    "adata_CD_1_raw.var_names_make_unique\n",
    "\n",
    "\n",
    "########\n",
    "\n",
    "\n",
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_CD_2 = './../filtered_data/CD_2/matrix.mtx'\n",
    "filename_genes_CD_2 = './../filtered_data/CD_2/genes.tsv'\n",
    "filename_barcodes_CD_2 = './../filtered_data/CD_2/barcodes.tsv'\n",
    "\n",
    "adata_CD_2 = sc.read(filename_data_CD_2, cache=True).transpose()\n",
    "adata_CD_2.var_names = np.genfromtxt(filename_genes_CD_2, dtype=str)[:, 1]\n",
    "adata_CD_2.obs_names = np.genfromtxt(filename_barcodes_CD_2, dtype=str)\n",
    "\n",
    "adata_CD_2.var_names_make_unique\n",
    "\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_CD_2 = './../raw_data/CD_2/matrix.mtx'\n",
    "filename_genes_CD_2 = './../raw_data/CD_2/genes.tsv'\n",
    "filename_barcodes_CD_2 = './../raw_data/CD_2/barcodes.tsv'\n",
    "\n",
    "adata_CD_2_raw = sc.read(filename_data_CD_2, cache=True).transpose()\n",
    "adata_CD_2_raw.var_names = np.genfromtxt(filename_genes_CD_2, dtype=str)[:, 1]\n",
    "adata_CD_2_raw.obs_names = np.genfromtxt(filename_barcodes_CD_2, dtype=str)\n",
    "\n",
    "\n",
    "adata_CD_2_raw.var_names_make_unique()\n",
    "\n",
    "\n",
    "############\n",
    "\n",
    "# filtered data (cell barcodes defined by Cellranger)\n",
    "filename_data_CD_3 = './../filtered_data/CD_3/matrix.mtx'\n",
    "filename_genes_CD_3 = './../filtered_data/CD_3/genes.tsv'\n",
    "filename_barcodes_CD_3 = './../filtered_data/CD_3/barcodes.tsv'\n",
    "\n",
    "adata_CD_3 = sc.read(filename_data_CD_3, cache=True).transpose()\n",
    "adata_CD_3.var_names = np.genfromtxt(filename_genes_CD_3, dtype=str)[:, 1]\n",
    "adata_CD_3.obs_names = np.genfromtxt(filename_barcodes_CD_3, dtype=str)\n",
    "\n",
    "adata_CD_3.var_names_make_unique()\n",
    "\n",
    "\n",
    "# raw data (all barcodes)\n",
    "filename_data_CD_3 = './../raw_data/CD_3/matrix.mtx'\n",
    "filename_genes_CD_3 = './../raw_data/CD_3/genes.tsv'\n",
    "filename_barcodes_CD_3 = './../raw_data/CD_3/barcodes.tsv'\n",
    "\n",
    "adata_CD_3_raw = sc.read(filename_data_CD_3, cache=True).transpose()\n",
    "adata_CD_3_raw.var_names = np.genfromtxt(filename_genes_CD_3, dtype=str)[:, 1]\n",
    "adata_CD_3_raw.obs_names = np.genfromtxt(filename_barcodes_CD_3, dtype=str)\n",
    "adata_CD_3_raw.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5083\n",
      "filtered out 732197 cells that have less than 3906.0 counts\n",
      "filtered out 731889 cells that have less than 1000 genes expressed\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "5407\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "5407\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "5407\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_CD_1=qc_metrics(adata_CD_1)\n",
    "\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_CD_1.obs['n_counts'])\n",
    "adata_CD_1_filt= cell_filter(adata_CD_1_raw,countmin)\n",
    "adata_CD_1_filt=qc_metrics(adata_CD_1_filt)\n",
    "\n",
    "adata_CD_1_filt = adata_CD_1_filt[adata_CD_1_filt.obs['percent_mito'] < 0.1, :]\n",
    "adata_CD_1_filt=qc_metrics(adata_CD_1_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4824\n",
      "filtered out 732456 cells that have less than 5008.0 counts\n",
      "filtered out 731994 cells that have less than 1000 genes expressed\n",
      "5298\n",
      "5298\n",
      "5298\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_CD_2=qc_metrics(adata_CD_2)\n",
    "\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_CD_2.obs['n_counts'])\n",
    "adata_CD_2_filt= cell_filter(adata_CD_2_raw,countmin)\n",
    "adata_CD_2_filt=qc_metrics(adata_CD_2_filt)\n",
    "\n",
    "adata_CD_2_filt = adata_CD_2_filt[adata_CD_2_filt.obs['percent_mito'] < 0.1, :]\n",
    "adata_CD_2_filt=qc_metrics(adata_CD_2_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4731\n",
      "filtered out 732549 cells that have less than 4502.0 counts\n",
      "filtered out 732022 cells that have less than 1000 genes expressed\n",
      "5271\n",
      "5271\n",
      "5271\n"
     ]
    }
   ],
   "source": [
    "### QC and cell filtering\n",
    "adata_CD_3=qc_metrics(adata_CD_3)\n",
    "\n",
    "# count threshold from Cellranger filtering (minimal total count)\n",
    "countmin = min(adata_CD_3.obs['n_counts'])\n",
    "adata_CD_3_filt= cell_filter(adata_CD_3_raw,countmin)\n",
    "adata_CD_3_filt=qc_metrics(adata_CD_3_filt)\n",
    "\n",
    "adata_CD_3_filt = adata_CD_3_filt[adata_CD_3_filt.obs['percent_mito'] < 0.1, :]\n",
    "adata_CD_3_filt=qc_metrics(adata_CD_3_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_CD_1_filt.var_names_make_unique()\n",
    "adata_CD_2_filt.var_names_make_unique()\n",
    "adata_CD_3_filt.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_all=adata_CD_1_filt.concatenate([adata_CD_2_filt, adata_CD_3_filt],\n",
    "                                       batch_key='sample',\n",
    "                                       batch_categories=['CD_1', 'CD_2', 'CD_3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 13033 genes that are detected in less than 20 cells\n",
      "15976\n"
     ]
    }
   ],
   "source": [
    "sc.pp.filter_genes(adata_all, min_cells=20)\n",
    "adata_all=qc_metrics(adata_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.write('./../data/control_only_new_filter_all_vers_1_4.h5',adata_all)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 0,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "nav_menu": {},
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "84.7159px",
    "width": "251.989px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 6,
   "toc_cell": false,
   "toc_section_display": "none",
   "toc_window_display": true
  },
  "toc_position": {
   "height": "689px",
   "left": "0px",
   "right": "1394.13px",
   "top": "110px",
   "width": "212px"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
